src/compute/src/sink/persist_sink.rs

```diff
@@ -285,17 +286,37 @@ where
         AsyncOperatorBuilder::new(format!("{} mint_batch_descriptions", operator_name), scope);
 
     let (mut output, output_stream) = mint_op.new_output();
+    let (mut data_output, data_output_stream) = mint_op.new_output();
 
+    // The description input and the data-passthrough input are both driven by this input, so
+    // they use a standard input connection.
     let mut desired_input = mint_op.new_input(desired_stream, Pipeline);
-    let mut persist_feedback_input =
-        mint_op.new_input_connection(persist_feedback_stream, Pipeline, vec![Antichain::new()]);
+
+    let mut persist_feedback_input = mint_op.new_input_connection(
+        persist_feedback_stream,
+        Pipeline,
+        // Neither output's capabilities should depend on the feedback input.
+        vec![Antichain::new(), Antichain::new()],
+    );
 
     let shutdown_button = mint_op.build(move |mut capabilities| async move {
+        // Non-active workers should just pass the data through.
         if !active_worker {
+            capabilities.clear();
+            while let Some(event) = desired_input.next_mut().await {
+                match event {
+                    Event::Data(cap, data) => {
+                        data_output.give_container(&cap, data).await;
+                    }
+                    Event::Progress(_) => {}
+                }
+            }
             return;
         }
 
-        let mut cap_set = CapabilitySet::from_elem(capabilities.pop().expect("missing capability"));
+        let mut cap_set = CapabilitySet::from_elem(capabilities.swap_remove(0));
+        // The data output should will use the data capabilities.
+        capabilities.pop();
```

Comment:
Maybe that's just me not being used to reading operator code, but I was confused about this part for a bit. I think this would be more readable:

```rust
        let [desc_cap, data_cap]: [_; 2] =
            capabilities.try_into().expect("one capability per output");
        let mut cap_set = CapabilitySet::from_elem(desc_cap);
        // The data output should will use the data capabilities.
        drop(data_cap);
```

Or even just:

```rust
        let [desc_cap, _data_cap]: [_; 2] =
            capabilities.try_into().expect("one capability per output");
        let mut cap_set = CapabilitySet::from_elem(desc_cap);
```

Reply:
much better!